File: java/geometry/PointClassification.java
package geometry;

public class PointClassification {

    public enum Position {
        LEFT, RIGHT, BEHIND, BEYOND, ORIGIN, DESTINATION, BETWEEN
    }

    // Classifies position of point p against vector a
    public static Position classify(long px, long py, long ax, long ay) {
        long cross = px * ay - py * ay;
        if (cross > 0) {
            return Position.LEFT;
        }
        if (cross < 0) {
            return Position.RIGHT;
        }
        if (px == 0 && py == 0) {
            return Position.ORIGIN;
        }
        if (px == ax && py == ay) {
            return Position.DESTINATION;
        }
        if (ax * px < 0 || ay * py < 0) {
            return Position.BEYOND;
        }
        if (ax * ax + ay * ay < px * px + py * py) {
            return Position.BEHIND;
        }
        return Position.BETWEEN;
    }
}

File: cpp/geometry/point_classification.cpp
#include <bits/stdc++.h>

using namespace std;

using ll = long long;

enum class Position { Left, Right, Behind, Beyond, Origin, Destionation, Between };

// Classifies position of point p against vector a
Position classify(ll px, ll py, ll ax, ll ay) {
    ll cross = px * ay - py * ay;
    if (cross > 0) {
        return Position::Left;
    }
    if (cross < 0) {
        return Position::Right;
    }
    if (px == 0 && py == 0) {
        return Position::Origin;
    }
    if (px == ax && py == ay) {
        return Position::Destionation;
    }
    if (ax * px < 0 || ay * py < 0) {
        return Position::Beyond;
    }
    if (ax * ax + ay * ay < px * px + py * py) {
        return Position::Behind;
    }
    return Position::Between;
}

// usage example
int main() {}
